File: General-DSA-Questions/Heap/4_kthSmallestElement.cpp
// Write a function to find the kth smallest element in a array using heap.

// Given an array of n elements, find the Kth smallest element using a heap.

// The basic approach is to sort the array and return the kth element.
// However, this approach has a time complexity of O(N log N) due to sorting.
// We can optimize this using a max-heap to find the kth smallest element in O(N log K) time.

// Approach:
// 1. Create a max-heap of size k.
// 2. Insert the first k elements into the max-heap.
// 3. For each remaining element in the array, compare it with the root of the max-heap.
// 4. If the current element is smaller than the root, replace the root with the current element and heapify.
// 5. The root of the max-heap will be the kth smallest element after processing all elements because we are maintaining the k smallest elements in the heap so that the root will be the largest among them.
// 6. Time Complexity: O(N log K) for finding the kth smallest element using a max-heap.
//      a. Building the max-heap with the first k elements takes O(k) time.
//      b. Processing each of the remaining (N − K) elements:
//          For each element, you may compare with the heap top (O(1)).
//          If it’s smaller, you:
//             Remove the max (O(log K))
//             Insert the new element (O(log K))
//          So, in the worst case, every element causes a remove+insert = O(log K).
//      Total processing time:
//          You process N − K elements, each costing O(log K):
//          (N − K) * O(log K) = O(N log K)
//      Thus, the overall time complexity is O(k + (N − K) log K). which simplifies to O(N log K).

// 7. Space Complexity: O(K) for storing the k smallest elements in the max-heap.

// This approach is efficient for large arrays and small values of k.

#include <iostream>
#include <queue>
using namespace std;

// Function to find the kth smallest element in an array using a max-heap
int kthSmallest(int arr[], int n, int k) {
    // Create a max-heap using priority queue
    priority_queue<int> maxHeap;

    // Insert the first k elements into the max-heap
    for (int i = 0; i < k; i++) {
        maxHeap.push(arr[i]);
    }

    // For each remaining element in the array
    for (int i = k; i < n; i++) {
        // If the current element is smaller than the root of the max-heap
        if (arr[i] < maxHeap.top()) {
            maxHeap.pop(); // Remove the largest element from the heap
            maxHeap.push(arr[i]); // Insert the current element into the heap
        }
    }

    // The root of the max-heap is the kth smallest element
    return maxHeap.top();
}

int main() {
  int arr[] = {7, 10, 4, 3, 20, 15};
  int n = sizeof(arr) / sizeof(arr[0]);
  int k = 3; // Find the 3rd smallest element
  cout << "The " << k << "rd smallest element is " << kthSmallest(arr, n, k) << endl;
  return 0;
}

// Output: The 3rd smallest element is 7

File: JS-DSA-Questions/Heap/4_kthSmallestElement.js
// Find the kth smallest element in an array using a heap.

// Sorting would be O(N log N). Using a MAX-heap of size k, we can do it in O(N log K):
// 1. Push the first k elements into a max-heap.
// 2. For each remaining element, if it is smaller than the heap's max (root), pop the max and push
//    it. The heap always holds the k smallest elements seen; its root is the kth smallest.
// Space: O(K).

// JavaScript note: JS has NO built-in priority queue, so we include a small MaxHeap class here
// (0-based array binary heap). In interviews you'd either write one like this or use a library.
class MaxHeap {
  constructor() {
    this.a = [];
  }
  size() {
    return this.a.length;
  }
  peek() {
    return this.a[0];
  }
  push(v) {
    this.a.push(v);
    let i = this.a.length - 1;
    while (i > 0) {
      const parent = Math.floor((i - 1) / 2);
      if (this.a[parent] < this.a[i]) {
        [this.a[parent], this.a[i]] = [this.a[i], this.a[parent]];
        i = parent;
      } else break;
    }
  }
  pop() {
    const top = this.a[0];
    const last = this.a.pop();
    if (this.a.length > 0) {
      this.a[0] = last;
      let i = 0;
      const n = this.a.length;
      while (true) {
        const l = 2 * i + 1;
        const r = 2 * i + 2;
        let largest = i;
        if (l < n && this.a[l] > this.a[largest]) largest = l;
        if (r < n && this.a[r] > this.a[largest]) largest = r;
        if (largest !== i) {
          [this.a[i], this.a[largest]] = [this.a[largest], this.a[i]];
          i = largest;
        } else break;
      }
    }
    return top;
  }
}

function kthSmallest(arr, k) {
  const maxHeap = new MaxHeap();

  for (let i = 0; i < k; i++) {
    maxHeap.push(arr[i]);
  }

  for (let i = k; i < arr.length; i++) {
    if (arr[i] < maxHeap.peek()) {
      maxHeap.pop();
      maxHeap.push(arr[i]);
    }
  }

  return maxHeap.peek();
}

const arr = [7, 10, 4, 3, 20, 15];
const k = 3;
console.log(`The ${k}rd smallest element is ${kthSmallest(arr, k)}`); // 7
